### PR TITLE
feat: mason search

### PR DESCRIPTION
**Pull request by Luckey-Elijah**
## Status

**READY**

## Description

_Add mason search feature as described in #379_

Adds Mason CLI command to search registered bricks on `brickhub.dev`.

## Type of Change

- [X] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

**Comment by Luckey-Elijah** [comment]
Here's a demo of `mason search` in action

https://user-images.githubusercontent.com/38171761/171409301-3e0053d7-b6ff-4484-95c9-2efe29bb03a2.mov

**Comment by Luckey-Elijah** [comment]
@felangel Could you review this PR?

Some things to note:
- uses the `api/v1/search` endpoint
- doesn't have a limit feature right now with `mason search term --limit=15` (I wanted to demonstrate at least POC)
- I defined a `Brick` model in `mason_api`, _but_ there exists another in `mason` (core package). This isn't ideal and didn't know if you have a preference on how to proceed

File: packages/mason_api/lib/src/mason_api.dart [hunk 1]
```diff
@@ -92,6 +101,20 @@ class MasonApi {
   /// The current user.
   User? get currentUser => _currentUser;
 
+  /// The host for
```
Comment (Luckey-Elijah) [inline packages/mason_api/lib/src/mason_api.dart:104]: Fix this doc comment

File: packages/mason_api/lib/src/models/brick.dart [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import 'package:json_annotation/json_annotation.dart';
+
+part 'brick.g.dart';
+
+/// {@template brick}
+/// Details of a brick from `brickhub.dev` registry
+/// {@endtemplate}
+@JsonSerializable(createToJson: false)
+class Brick {
```
Comment (Luckey-Elijah) [inline packages/mason_api/lib/src/models/brick.dart:9]: There is a brick already defined in `mason` (core package)

File: packages/mason_api/lib/src/models/brick.dart [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  /// {@macro brick}
+  const Brick({
```
Comment (Luckey-Elijah) [inline packages/mason_api/lib/src/models/brick.dart:11]: Needs `publisher` field

File: packages/mason_cli/lib/src/commands/search.dart [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      final results = await _masonApi.search(query: query);
+      if (results.isEmpty) {
+        searchProgress.complete('No bricks found.');
+      } else {
+        searchProgress.complete(
+          'Found ${results.length} brick${results.length > 1 ? 's' : ''}.',
+        );
+
+        for (final brick in results) {
+          logger
```
Comment (Luckey-Elijah) [inline packages/mason_cli/lib/src/commands/search.dart:37]: Should this include the publisher as well?

**Comment by felangel** [comment]
@Luckey-Elijah thanks! I'll take a look tomorrow 👍

File: packages/mason_api/lib/src/models/brick.dart (hunk 2 above)
Comment (Luckey-Elijah) [reply packages/mason_api/lib/src/models/brick.dart:9]: That's a great idea

**Comment by Luckey-Elijah** [comment]
Made the code refactor for `Brick` -> `BrickSearchResult`. I think this is better and more clear for what the model represents.

Additionally CLI results appears like:
Screen Shot 2022-06-02 at 10 33 48 PM

**Comment by felangel** [comment]
@Luckey-Elijah I'm making some minor adjustments and will merge this shortly! Thanks so much for the amazing contribution 💙

**Review by felangel — APPROVED** [review]
LGTM 🎉 💙

**Pull request merged**